From 74f87587bac645cc38ffdd726c0fefbbdf0092c6 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Marek=20Marczykowski-G=C3=B3recki?=
 <marmarek@invisiblethingslab.com>
Date: Tue, 3 Feb 2015 05:31:38 +0100
Subject: [PATCH] xenconfig: parse vif/backend attribute, tests
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit
Organization: Invisible Things Lab
Cc: Marek Marczykowski-Górecki <marmarek@invisiblethingslab.com>

The support for backend domain name in xenParseSxprNets is mostly to add
some tests for generating XML with vif connected to a driver domain.

Signed-off-by: Marek Marczykowski-Górecki <marmarek@invisiblethingslab.com>
---
 src/xenconfig/xen_sxpr.c                           |  4 +++
 .../sexpr2xml-bridge-driver-domain.sexpr           |  9 +++++
 .../sexpr2xml-bridge-driver-domain.xml             | 38 ++++++++++++++++++++++
 tests/sexpr2xmltest.c                              |  1 +
 4 files changed, 52 insertions(+)
 create mode 100644 tests/sexpr2xmldata/sexpr2xml-bridge-driver-domain.sexpr
 create mode 100644 tests/sexpr2xmldata/sexpr2xml-bridge-driver-domain.xml

diff --git a/src/xenconfig/xen_sxpr.c b/src/xenconfig/xen_sxpr.c
index e868c05695..5e1eaaa24c 100644
--- a/src/xenconfig/xen_sxpr.c
+++ b/src/xenconfig/xen_sxpr.c
@@ -625,6 +625,10 @@ xenParseSxprNets(virDomainDefPtr def,
                     goto cleanup;
             }
 
+            tmp = sexpr_node(node, "device/vif/backend");
+            if (tmp && STRNEQ(tmp, "0") &&
+                    VIR_STRDUP(net->domain_name, tmp) < 0)
+                goto cleanup;
             tmp = sexpr_node(node, "device/vif/vifname");
             /* If vifname is specified in xend config, include it in net
              * definition regardless of domain state.  If vifname is not
diff --git a/tests/sexpr2xmldata/sexpr2xml-bridge-driver-domain.sexpr b/tests/sexpr2xmldata/sexpr2xml-bridge-driver-domain.sexpr
new file mode 100644
index 0000000000..b44876aaec
--- /dev/null
+++ b/tests/sexpr2xmldata/sexpr2xml-bridge-driver-domain.sexpr
@@ -0,0 +1,9 @@
+(domain (domid 6)(name 'pvtest')(memory 420)(maxmem 420)(vcpus 2)\
+(uuid '596a5d2171f48fb2e068e2386a5c413e')(on_poweroff 'destroy')\
+(on_reboot 'destroy')(on_crash 'destroy')\
+(image (linux (kernel '/var/lib/xen/vmlinuz.2Dn2YT')\
+(ramdisk '/var/lib/xen/initrd.img.0u-Vhq')\
+(args ' method=http://download.fedora.devel.redhat.com/pub/fedora/linux/\
+core/test/5.91/x86_64/os  ')))(device (vbd (dev 'xvda')\
+(uname 'file:/root/some.img')(mode 'w')))\
+(device (vif (mac '00:11:22:33:44:55')(bridge 'xenbr2')(script 'vif-bridge')(backend 'netvm'))))
diff --git a/tests/sexpr2xmldata/sexpr2xml-bridge-driver-domain.xml b/tests/sexpr2xmldata/sexpr2xml-bridge-driver-domain.xml
new file mode 100644
index 0000000000..60815a8111
--- /dev/null
+++ b/tests/sexpr2xmldata/sexpr2xml-bridge-driver-domain.xml
@@ -0,0 +1,38 @@
+<domain type='xen' id='6'>
+  <name>pvtest</name>
+  <uuid>596a5d21-71f4-8fb2-e068-e2386a5c413e</uuid>
+  <memory unit='KiB'>430080</memory>
+  <currentMemory unit='KiB'>430080</currentMemory>
+  <vcpu placement='static'>2</vcpu>
+  <os>
+    <type>linux</type>
+    <kernel>/var/lib/xen/vmlinuz.2Dn2YT</kernel>
+    <initrd>/var/lib/xen/initrd.img.0u-Vhq</initrd>
+    <cmdline> method=http://download.fedora.devel.redhat.com/pub/fedora/linux/core/test/5.91/x86_64/os  </cmdline>
+  </os>
+  <clock offset='utc' adjustment='reset'/>
+  <on_poweroff>destroy</on_poweroff>
+  <on_reboot>destroy</on_reboot>
+  <on_crash>destroy</on_crash>
+  <devices>
+    <disk type='file' device='disk'>
+      <driver name='file'/>
+      <source file='/root/some.img'/>
+      <backingStore/>
+      <target dev='xvda' bus='xen'/>
+    </disk>
+    <interface type='bridge'>
+      <mac address='00:11:22:33:44:55'/>
+      <source bridge='xenbr2'/>
+      <script path='vif-bridge'/>
+      <backenddomain name='netvm'/>
+      <target dev='vif6.0'/>
+    </interface>
+    <console type='pty'>
+      <target type='xen' port='0'/>
+    </console>
+    <input type='mouse' bus='xen'/>
+    <input type='keyboard' bus='xen'/>
+    <memballoon model='xen'/>
+  </devices>
+</domain>
diff --git a/tests/sexpr2xmltest.c b/tests/sexpr2xmltest.c
index efe64110d8..ebc3358b22 100644
--- a/tests/sexpr2xmltest.c
+++ b/tests/sexpr2xmltest.c
@@ -127,6 +127,7 @@ mymain(void)
     DO_TEST("net-bridged", "net-bridged");
     DO_TEST("net-e1000", "net-e1000");
     DO_TEST("bridge-ipaddr", "bridge-ipaddr");
+    DO_TEST("bridge-driver-domain", "bridge-driver-domain");
     DO_TEST("no-source-cdrom", "no-source-cdrom");
     DO_TEST("pv-localtime", "pv-localtime");
     DO_TEST("pci-devs", "pci-devs");
-- 
2.13.6

